### ISSUE TITLE
Android 8.0 Oreo and firebird

### ISSUE THREAD

**Issue by Jo900** [nspire-emus/firebird#115]
Today is started the rollout of Oreo Android 8.0, for my device. 
I not received yet, i think will be tomorrow.
So is not a issue, is just a question... firebird1.4 can work good on oreo ? Somebody already tried ?
Wish you all a great 2018 !

**Comment by Vogtinator** [comment]
I don't know - I only have devices with Android 5.0 and 4.4.2.

The apk from https://github.com/nspire-emus/firebird/issues/111#issuecomment-350049067 should work fine if you don't have an external SD card.

**Comment by Jo900** [comment]
Not tried yet Oreo, i'm waiting for notification. 
I have sd card, 
emulator files are into main memory :)

**Comment by Jo900** [comment]
ok, works well on Android 8.0.0 Oreo.
There are only some minor visual font missing
https://s17.postimg.org/421fua1nj/image.png

and press button/key most of the time stay stuck and i have to press again to release it, a bit annoying.

is possible to have soft key vibration ? would help the feeling as real "click"...
thanks

**Comment by Jo900** [comment]
another little suggestion:
at key press would be nice have little "tick-click" sound that simulate real sound of calculator keyboard.

**Comment by Vogtinator** [comment]
I already thought about vibration as feedback, but it's not that easy to implement without going into platform-native code which I have no idea of (Android and possibly iOS).

Pull-requests welcome, it's not that hard if you know the APIs.

**Comment by Jo900** [comment]
a "click" sound is more useful than a vibration as feedback, is possible to have just the sound as 
key feedback ? :)
Another suggestion, would be much fast to use with the gui 1:1 on the screen, i know is impossible to keep real proportions (can be with black stripes on sides or more short and compact). Swipe down and up everytime, with key stuck at every swipe and no sound make very probable type errors.. and is not so comfortable to use :) (is the app i use more often when i study, and if very useful since i don't need to bring with me the "real" nspire).
Thanks

**Comment by Vogtinator** [comment]
> a "click" sound is more useful than a vibration as feedback, is possible to have just the sound as
key feedback ? :)

Sure, can you suggest a sound with a compatible license?

> would be much fast to use with the gui 1:1 on the screen

On a tablet maybe, there's special support for that in landscape mode. On a phone, probably not. The keys would be too small to use.

> with key stuck at every swipe

I don't get that here.

**Comment by Jo900** [comment]
When you swipe up and down the keyboard, the key you touch stay press down, and i have to press again to release it. This or i have to swipe on empty space between key, and when i'm in hurry with calculation is an element of distraction.

What you mean compatible license for sound ? The typical click sound of android (when you press any icon would be great) Or something similar with free license, or i can record the real(so cool) click sound of nspire and sent to you (i have good mic and mixer).

**Comment by Vogtinator** [comment]
> When you swipe up and down the keyboard, the key you touch stay press down, and i have to press again to release it. This or i have to swipe on empty space between key, and when i'm in hurry with calculation is an element of distraction.

I see - QML's MultiPointTouchArea which is needed for multitouch doesn't have "preventStealing" set, so it's impossible to disable scrolling on buttons like it's done on desktop.
Maybe I can find a workaround.

> Or something similar with free license, or i can record the real(so cool) click sound of nspire and sent to you (i have good mic and mixer).

Obviously it needs to work on all platforms, so we need our own sound file, whatever it is.

**Comment by Vogtinator** [comment]
This should work, scrolling now releases the keys. It might be too sensitive though.

[android-build-debug.zip](https://github.com/nspire-emus/firebird/files/1777653/android-build-debug.zip)

**Comment by Jo900** [comment]
I tested, work just the key release, scrolling not work when key is pressed, work only between keys.
And is back the crash on exit :)))
I think the best release so far was the one before this, with the key sound is just perfect. Not had time yet to make audio record, maybe this weekend. :))
Thanks

**Comment by Vogtinator** [comment]
> I tested, work just the key release, scrolling not work when key is pressed, work only between keys.

Damn it, that's a bug in Qt then :-(

> And is back the crash on exit :)))

Possible, I might have built it using the wrong Qt version.

> I think the best release so far was the one before this

1.2? That didn't have multi-touch support, which means you could only press one button at a time.

**Comment by Jo900** [comment]
no, is not the 1.2, is the 1.4-dev that you build with the right version of Qt to avoid crash on exit :))

So, to make more comfortable for intensive use (as i do at engineering faculty) i think would be great the key sound and if possible a new optional key layout (just proportions) to avoid scrolling... example a menu where you can choose between scrolling or fixed keyboard

**Comment by adriweb** [comment]
I went looking for a keypress sound, maybe this one is okay enough:
https://www.zapsplat.com/music/multimedia-button-click-13/ ? License seems okay too (basically needs attribution and no modification, but no re-share might be a bit hard with the repo though)

These might be interesting as well (CC licenses, possibly better than the one above):
- https://freesound.org/people/Jagadamba/sounds/254286/ (for keydown?)
- https://freesound.org/people/Qat/sounds/108336/ (for keyup?)
- https://freesound.org/people/potentjello/sounds/194056/

You can find more here https://freesound.org/search/?a_tag=1&g=1&f=duration%3A%5B0+TO+1%5D&q=button&s=duration+asc&advanced=1

**Comment by Jo900** [comment]
the first sound button click is perfect :)

**Comment by Vogtinator** [comment]
> I went looking for a keypress sound, maybe this one is okay enough:
https://www.zapsplat.com/music/multimedia-button-click-13/ ? License seems okay too (basically needs attribution and no modification, but no re-share might be a bit hard with the repo though)

Nope, the license is not compatible.

Attribution is ok, but non commercial and no redistribution aren't compatible.

> BTW, this is probably interesting, implementation-wise: https://doc.qt.io/qt-5/qml-qtmultimedia-soundeffect.html The example being what we probably want.

It goes into the right direction, but we support Qt 5.6+. http://doc.qt.io/qt-5/qml-qtmultimedia-audio.html is usable.

**Comment by adriweb** [comment]
Yeah the license is problematic, I guess.
The CC-licensed freesound.org ones should be good though?

**Comment by Vogtinator** [comment]
Yup.

This might need some internal changes to handle the case correctly when you're typing with the keyboard and using the touchscreen at the same time. Hm.

**Comment by Jo900** [comment]
Any news about 1.5 version with keypress sound ?
Would be possible re-arrange key in querty layout ? (nice trick and very comfortable to write).
Emulator is great, there is only need of few daily use impro.
p.s. don't forget that is still present the bug on qt that cause crash on exit, keep in mind for upcoming releases.

Thanks.

**Comment by Vogtinator** [comment]
I'm not fully convinced yet which sound to use.

There is a branch which enables the use of the android keyboard to type input. I'm not sure how well that works though.

**Comment by Jo900** [comment]
Any sound that help with some kind of key feedback, using this emulator on smartphone has only few downside, all related to daily comfortable usability. Very often i make a type mistake and i not even realize, cause i not press properly (just key missing). That's sound click would help a lot ! 

Another usability impro trick can be re-arrange key in querty (not needed android keyboard), just re-arrange key by editing the skin image and key links.

About crash bug on exit, i tried latest develop build uploaded on "multiple file upload" thread. And is still present the very annoying crash bug.

Emulator is almost perfect, work properly, and as i'm learning everyday how use more function and program of my calculator it start to be one of the best app on android smartphone. I'm an engineer, and happen often i have to make some fast calculations. 

p.s. i don't know if this project support donations, even a small one, just for a beer. :)))

**Comment by zhenniqi** [comment]
May i ask if there is haptic feedback exist in latest 1.6 version,if not,can it be added?thanks.

**Comment by Vogtinator** [comment]
No, it's not in 1.6 yet.

**Comment by davidye-git** [comment]
Will haptic feedback ever get added to the android build? As other people mentioned, it would help with improving accuracy of key inputs.

**Comment by Vogtinator** [comment]
#362 implements this, you can download a test APK from https://github.com/nspire-emus/firebird/actions/runs/13146842156.
